### PR TITLE
ISSUE 414 fix

### LINKED ISSUE #414: Make CSV report better prepared for further processing
For example in LibreOffice Calc - currently when one wants to generate charts based on the output, a lot of processing/filtering is required. The most important columns are throughput and response time.
All data should be in the CSV in a way that makes it easy to aggregate the data from multiple configurations, and then do additional analysis. CSV = comma seperated too! (Not semi-colons!)

### PR DESCRIPTION
**Pull request by zhostasa**
Minor formatting
Changed CSV separator to comma
Added RowComparator and ColumnComparator to sort rows and columns,
Added column order optional argument to CSV reporter

resolves #414

### CODE REVIEW

**Comment by jmarkos** [comment]
@mgencur Could you look at this one please? The issue was pretty vague so I cannot determine if this PR actually fixes it or not.

**Comment by mgencur** [comment]
OK. Will do.

File: reporters/reporter-default/src/main/java/org/radargun/reporting/csv/CsvReporter.java [hunk 1]
```diff
@@ -287,4 +325,95 @@ private double round(double number, int places) {
       return bigDecimal.doubleValue();
    }
 
+   /**
+    * Comparator for sorting rows based on column value
+    * 
+    * @author zhostasa
```
Comment (jmarkos) [inline reporters/reporter-default/src/main/java/org/radargun/reporting/csv/CsvReporter.java]: It was decided in https://github.com/radargun/radargun/pull/385 to remove the author tags, so I think no new ones should be added.

File: reporters/reporter-default/src/main/java/org/radargun/reporting/csv/CsvReporter.java [hunk 2]
```diff
@@ -41,14 +54,18 @@
...
-   private String separator = ";";
+   @Property(doc = "Separator of columns in the CSV file. Default is ','")
+   private String separator = ",";
 
    @Property(doc = "Compute aggregated statistics from all nodes. Default is true.")
    private boolean computeTotal = true;
 
    @Property(doc = "Compute response times at certain percentiles. Default is 95% and 99%.")
-   protected double[] percentiles = new double[] {95d, 99d};
+   protected double[] percentiles = new double[] { 95d, 99d };
```
Comment (jmarkos) [inline reporters/reporter-default/src/main/java/org/radargun/reporting/csv/CsvReporter.java]: I'd prefer to not change such things unless checkstyle complains.

File: reporters/reporter-default/src/main/java/org/radargun/reporting/csv/CsvReporter.java [hunk 3]
```diff
@@ -66,7 +83,7 @@ private void reportTest(Report report, Report.Test test) {
          fileWriter = prepareOutputFile(report, test.name, "");
          int it = 0;
          Set<String> columns = new TreeSet<String>();
-         ArrayList<Map<String, String>> rows = new ArrayList();
+         ArrayList<Map<String, String>> rows = new ArrayList<Map<String, String>>();
```
Comment (jmarkos) [inline reporters/reporter-default/src/main/java/org/radargun/reporting/csv/CsvReporter.java]: Same here.

File: reporters/reporter-default/src/main/java/org/radargun/reporting/csv/CsvReporter.java [hunk 4]
```diff
@@ -132,10 +151,11 @@ private FileWriter prepareOutputFile(Report report, String prefix, String suffix
       return new FileWriter(outputFile);
    }
 
-   private Statistics processRow(int it, Set<String> columns, List<Map<String, String>> rows, Map.Entry<Integer, List<Statistics>> slaveStats) {
+   private Statistics processRow(int it, Set<String> columns, List<Map<String, String>> rows,
+         Map.Entry<Integer, List<Statistics>> slaveStats) {
       // this reporter is merging statistics from all threads on each node
       Statistics summary = slaveStats.getValue().stream().filter(o -> o != null).reduce(Statistics.MERGE)
-         .orElseThrow(() -> new IllegalStateException("No statistics!"));
+            .orElseThrow(() -> new IllegalStateException("No statistics!"));
```
Comment (jmarkos) [inline reporters/reporter-default/src/main/java/org/radargun/reporting/csv/CsvReporter.java]: And here.

File: reporters/reporter-default/src/main/java/org/radargun/reporting/csv/CsvReporter.java [hunk 5]
```diff
@@ -150,11 +170,12 @@ private Statistics processRow(int it, Set<String> columns, List<Map<String, Stri
...
 
-   private void addRepresentations(Statistics summary, Map<String, String> rowData, Statistics statistics, String operationName) {
-      //Map.Entry<Integer, List<Statistics>> slaveStats = ; Map.Entry<String, OperationStats> os = ;
+   private void addRepresentations(Statistics summary, Map<String, String> rowData, Statistics statistics,
+         String operationName) {
       DefaultOutcome defaultOutcome = statistics.getRepresentation(operationName, DefaultOutcome.class);
       if (defaultOutcome != null) {
-         if (defaultOutcome.requests == 0) return;
+         if (defaultOutcome.requests == 0)
+            return;
```
Comment (jmarkos) [inline reporters/reporter-default/src/main/java/org/radargun/reporting/csv/CsvReporter.java]: One line if's are frequently used, not sure if we want to start rewriting them :/ Not saying I agree with such coding style.

File: reporters/reporter-default/src/main/java/org/radargun/reporting/csv/CsvReporter.java (hunk 1 above)
Comment (zhostasa) [inline reporters/reporter-default/src/main/java/org/radargun/reporting/csv/CsvReporter.java]: I just went over that discussion...it seems far from "decided"...
Anyway, I do not really care one way or the other

File: reporters/reporter-default/src/main/java/org/radargun/reporting/csv/CsvReporter.java (hunk 5 above)
Comment (zhostasa) [inline reporters/reporter-default/src/main/java/org/radargun/reporting/csv/CsvReporter.java]: I am saying I don't, but I will not be starting a flamewar over that (that is where these discussions generally end)

File: reporters/reporter-default/src/main/java/org/radargun/reporting/csv/CsvReporter.java (hunk 3 above)
Comment (zhostasa) [inline reporters/reporter-default/src/main/java/org/radargun/reporting/csv/CsvReporter.java]: This should be generating compile time warning and, more importantly, annoys my eclipse. I am pretty sure this should be done...come on...

Comment (mgencur) [inline reporters/reporter-default/src/main/java/org/radargun/reporting/csv/CsvReporter.java]: Yes. Using just ArrayList() produces a warning. Another way to fix this is the diamond operator:
ArrayList<Map<String, String>> rows = new ArrayList<>();

Comment (alanfx) [inline reporters/reporter-default/src/main/java/org/radargun/reporting/csv/CsvReporter.java]: Don't you just need this?

```
ArrayList<Map<String, String>> rows = new ArrayList<>();
```

Comment (mgencur) [inline reporters/reporter-default/src/main/java/org/radargun/reporting/csv/CsvReporter.java]: I was first! :)

Comment (alanfx) [inline reporters/reporter-default/src/main/java/org/radargun/reporting/csv/CsvReporter.java]: My network is too slow! :-)

Comment (zhostasa) [inline reporters/reporter-default/src/main/java/org/radargun/reporting/csv/CsvReporter.java]: Ok

File: reporters/reporter-default/src/main/java/org/radargun/reporting/csv/CsvReporter.java (hunk 1 above)
Comment (rvansa) [inline reporters/reporter-default/src/main/java/org/radargun/reporting/csv/CsvReporter.java]: Yep, the PR was never integrated - though my opinion is that the tags get obsolete too often and VCS is much better in preserving the authorship information - that's why I was removing my own tags only in the end.
Whoever wants to put their name there, can do (I would only ask for full name, not just username).

**Comment by zhostasa** [comment]
I have addressed most of the issues, feel free to go over it again

**Comment by mgencur** [comment]
Looks good to me except the commit message. "Minor formatting" is not really descriptive:) Let me test this and fix the commit message when pushing the changes to master.

**Comment by mgencur** [comment]
Hmm. Running a test with these changes results in empty CSV files and empty HTML tables. There's no data. Let me check if this happens with master branch without this PR.

**Comment by mgencur** [comment]
I got the benchmark to run successfully. The columns and rows seem to be re-ordered in the right way. However, I still see Gross/Net in the columns. We removed this in the tables in HTML report but it is still here in the CSV reports. We should align them: Throughput and Throughput w/ errors. Sorry that I didn't tell you earlier.

**Comment by zhostasa** [comment]
Done

**Comment by jmarkos** [comment]
Merged, thanks!

**Pull request closed without merging**